### PR TITLE
include withdraw/redeem sandwich testing

### PR DESCRIPTION
**Pull request by mcclurejt**
also fully implements idle liquidity maintenance for the vault.

### CODE REVIEW

File: contracts/EverlongStrategy.sol [hunk 1]
```diff
@@ -142,17 +142,38 @@ contract EverlongStrategy is BaseStrategy {
...
-        // Close all matured positions (if any).
-        uint256 output = _closeMaturedPositions(0);
-
-        // Close immature positions if additional funds need to be freed.
-        if (_amount > output) {
-            _closePositions(_amount - output);
+        // Calculate the current `totalAssets` and retrieve the previous value.
+        uint256 idle = asset.balanceOf(address(this));
+        uint256 currentTotalAssets = calculatePortfolioValue() + idle;
+        uint256 previousTotalAssets = TokenizedStrategy.totalAssets();
```
Comment (jalextowle) [inline contracts/EverlongStrategy.sol:155]: Is this updated when deploying and freeing funds? I'm just wondering if there is any way that this value and the current total assets are measuring different things.

File: contracts/EverlongStrategy.sol [hunk 2]
```diff
@@ -142,17 +142,38 @@ contract EverlongStrategy is BaseStrategy {
...
+            // Calculate the withdrawer's proportion of losses.
+            //
+            // It's important to use the TOTAL withdrawal amount, not just the
+            // amount being freed, when calculating the withdrawer's share.
+            //
+            //     totalWithdrawalAmount = _amount + idle
+            //
+            uint256 loss = previousTotalAssets - currentTotalAssets;
+            uint256 proportionalLoss = (loss).mulDivDown(
+                _amount + idle,
```
Comment (jalextowle) [inline contracts/EverlongStrategy.sol:169]: Just to make sure I understand this logic, I'll summarize my understanding of this calculation.

The gist is simple ofc, we just want to find the loss that is appropriate for this trader to pay. We just scale the total loss to the portion of funds that are actually being moved. My questions have to do with why we use `_amount + idle` and `previousTotalAssets` to calculate this proportion. 

`_amount + idle` is the amount of funds being withdrawn (`_freeFunds` is called during a withdrawal, and we know that idle is prioritized). This makes sense. Next up is `previousTotalAssets`. This was initially confusing to me, but it makes sense considering that `totalAssets` is used to compute the vault share price. 

IMO it would be good to make a comment justifying why we use `previousTotalAssets`

**Review by jalextowle — APPROVED** [review]
LGTM once comments are resolved

File: contracts/EverlongStrategy.sol (hunk 1 above)
Comment (mcclurejt) [reply contracts/EverlongStrategy.sol:155]: It's increased by the deposit amount when a vault adds debt to the strategy, and it's decreased by the amount freed  + any idle liquidity during `free_funds`.

I ~think~ that both of those changes would be reflected similarly for the current and previous `totalAssets`.

On deposit the previous `totalAssets` would be increased by the deposit amount and the current would be increased by the deposit amount less hyperdrive fees.

On `free_funds`, both previous and current `totalAssets` would be decremented by the real amount of assets transferred out of the strategy (NOT the amount requested to be freed afaik) so they should be equally affected.

Seems kind of weird but afaik it's an apples to apples comparison.

File: contracts/EverlongStrategy.sol (hunk 2 above)
Comment (mcclurejt) [reply contracts/EverlongStrategy.sol:169]: Yeah so this one definitely put my head in a blender a bit....

The redeemer's proportional share of the portfolio losses is as follows (assuming losses have occurred) :
```
redeemer_loss_share = total_portfolio_losses * value_of_longs_closed / (previous_portfolio_value - current_portfolio_value)
```

Unfortunately `totalAssets` is a combination of portfolio value and idle assets, and we don't know the proportion of each when looking at the stored value.... hugely problematic if idle liquidity varied dramatically, but it doesn't.

For our case, the strategy will almost always have zero idle (except right after a redemption due to the `partialPositionClosureBuffer`), and the maximum amount of idle it can have is just under Hyperdrive's `minimumTransactionAmount`. 

Since idle liquidity won't vary greatly and has little effect on `totalAssets`, I think it's safe for us to just compare `totalAssets` to determine bond portfolio losses.

**Comment by mcclurejt** [comment]
includes changes from https://github.com/delvtech/everlong/pull/16
